### ISSUE TITLE
Not finding sections when line is led with a space.

### ISSUE THREAD

**Issue by mlbelobraydi** [kinverarity1/lasio#359]
Receiving these warnings on file.
Header section Well regexp=~W was not found.
Header section Curves regexp=~C was not found.
Header section Parameter regexp=~P was not found.
No data section (regexp='~A') found

All sections exist, but have a space offset at the beginning of the line before ~*.
Result is that no sections are found and the data section is skipped.

I have many examples from COGCC that can be looked at, please let me know if you want more than the examples listed below.

Files: 
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2584534
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2589798
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2594183
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2578789
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2580893
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2247295
http://ogccweblink.state.co.us/DownloadDocumentPDF.aspx?DocumentId=2591655

**Comment by dcslagel** [comment]
Hi @mlbelobraydi,

I ran the first file in your list using Lasio 0.25.1  and everything seems to work.  Could you add more steps to reproduce the issue?  Is this error seen in a script, or in a jupyter workbook? 

Here are the steps I took on the command line:
```bash
$ python
>>> import lasio
>>> from lasio import read
>>> mylas = read('700016557.LAS')
>>> mylas.sections['Well']['STRT']
HeaderItem(mnemonic="STRT", unit="F", value="4986.0", descr="")
>>>  mylas.sections['Curves'].keys()  
['DEPT', 'CFTR1', 'CFTR2', 'NRMGR', 'NNILD', 'NNNPHI', 'NNRHOB', 'NNDPHI', 'CCL'] 
```

Thanks,
DC

**Comment by mlbelobraydi** [comment]
The attached .txt is actually a .ipynb notebook to show you what I am attempting to do.  If you have an alternative method please let me know. I would prefer not to write each file to disk, but to keep them in memory.
Thanks,
Matt

[Example for lasio.txt](https://github.com/kinverarity1/lasio/files/4814818/Example.for.lasio.txt)

**Comment by dcslagel** [comment]
Okay, yes, now I'm seeing the issue when trying to lasio.read(url)  instead of downloading the file first.  My initial guess is that the full url content is not completed reading before parsing is attempted.  Checking further....
DC

**Comment by dcslagel** [comment]
Triage findings:

This seems to be an encoding issue..

The issue looks to be that parts of these files contain `^M` (carriage-return) characters. 

When read from a url the encoding is set to `utf-8`. This correctly parses all the lines without `^M` characters.  It leaves the lines with `^M` as one big line and the section headers `~C`, `~P` and `~A` are then somewhere in the middle of  'multi-line' lines.  So those headers aren't found and parsed.   

However when the file is downloaded and saved to disk, Lasio, sees it's encoding as ASCII and parses the `^M` characters as line separators so the `~C`, `~P`, and `~A` sections are found and properly parsed.  

Looking further at why the encoding are set differently for the 2 situations.
DC

**Comment by dcslagel** [comment]
@mlbelobraydi,

Which python version are you running this with?

Thanks,
DC

**Comment by mlbelobraydi** [comment]
Anaconda 4.8.3
Python 3.7.7

**Comment by dcslagel** [comment]
@mlbelobraydi , @kinverarity1 ,

Okay pull-request https://github.com/kinverarity1/lasio/pull/360 `Use universal-newline when reading urllib content` should fix this issue.


This solution uses universal-newline in StringIO for reading url responses in url lib
This handles cases where a file has mixed '\r\n' and '\r' line endings.

References:
https://docs.python.org/3/library/io.html#io.StringIO
https://docs.python.org/3/library/io.html#io.TextIOWrapper

**Comment by kinverarity1** [comment]
Whoops sorry, didn't realise merging the pr would close this!

**Comment by dcslagel** [comment]
@mlbelobraydi,

The attached file is a temporary revision of your Example for lasio.txt (ipynb jupyter notebook) example that should run with your current lasso (lasio 0.25.1?).  This is a hack that overwrites lasio.reader.open_files() with the current master version with the fix + some additional scope handling.  Once the new version of Lasio is available on PyPy, update to the it and remove the cell with the hack.

Let us know if this works.

Thanks,
DC

[TempHack-Example.for.lasio.txt](https://github.com/kinverarity1/lasio/files/4831589/TempHack-Example.for.lasio.txt)

**Comment by dcslagel** [comment]
Test case is added in pull request https://github.com/kinverarity1/lasio/pull/361

**Comment by mlbelobraydi** [comment]
This now works for me in the notebook. I've added the script to my main program and it is having some success. After it finishes, I'll look into the files that failed.
Thanks, Matt

**Comment by dcslagel** [comment]
Okay, great!, Thanks for the update.

**Comment by kinverarity1** [comment]
I'll close this issue as it appears to be fixed - @mlbelobraydi, please feel free to re-open if you have further problems with it.

**Comment by kinverarity1** [comment]
FYI this issue should have been fixed with the release [v0.26](https://github.com/kinverarity1/lasio/releases/tag/v0.26)
